bioconda_utils/docker_utils.py

```diff
@@ -0,0 +1,325 @@
+        self.tag = tag
+        self.verbose = verbose
+        self.requirements = requirements
+        uid = os.getuid()
+        usr = pwd.getpwuid(uid)
+        self.user_info = dict(
+            uid=uid,
+            gid=usr.pw_gid,
+            groupname=grp.getgrgid(usr.pw_gid).gr_name,
+            username=usr.pw_name)
+        self.container_recipe = container_recipe
+        self.container_staging = container_staging
+
+        # Don't import until here to allow module import if docker not
+        # installed.
+        try:
+            from docker import Client as DockerClient
+        except ImportError:
+            raise ImportError("docker-py module must be installed")
+        self.docker = DockerClient(base_url=base_url)
+        self.host_conda_bld = get_host_conda_bld()
+        self._build_container()
+
+    def _build_container(self):
+        """
+        Builds a new image with requirements installed.
+        """
+
+        # Create a temporary build directory since we'll be copying the
+        # requirements file over
+        build_dir = tempfile.mkdtemp()
+
+        logger.debug('Building image "%s" from %s', self.tag, build_dir)
+        with open(os.path.join(build_dir, 'requirements.txt'), 'w') as fout:
+            if self.requirements:
+                fout.write(open(self.requirements).read())
+            else:
+                fout.write(open(pkg_resources.resource_filename(
+                    'bioconda_utils', 'bioconda_utils-requirements.txt')).read())
+
+        with open(os.path.join(build_dir, "Dockerfile"), 'w') as fout:
+            fout.write(dedent(
+                """\
+                FROM {self.image}
+                COPY requirements.txt /tmp/requirements.txt
+                RUN /opt/conda/bin/conda install --file /tmp/requirements.txt
+                """.format(**locals())))
+
+        logger.debug('Dockerfile:\n' + open(fout.name).read())
+
+        # TODO: need a better way of detecting build failure
+        response =self.docker.build(path=build_dir, rm=True, tag=self.tag)
+        response = ''.join(i.decode() for i in response)
+        if 'errorDetail' in response:
+            raise ValueError(''.join(response))
+
+        self._build = response
+        logger.info('Built docker image tag=%s', self.tag)
+        logger.debug(self._build)
+        shutil.rmtree(build_dir)
+        return self._build
+
+    def _write_build_script(self):
+        """
+        Writes a script to do the recipe building and copying over to mounted
+        host dir.
+
+        Filled in using currently-configured self.container_recipe and
+        self.container_staging.
+
+        Expects `HOST_USER` and `CONDA_BUILD_ARGS` to be passed in as
+        environmental variables.
+        """
+        build_dir = tempfile.mkdtemp()
+        with open(os.path.join(build_dir, 'build_script.bash'), 'w') as fout:
+            fout.write(dedent(
+                """
+                #!/bin/bash
```

Comment:
Can we put this on top of this file to and make it more obvious and easy to change.